### ISSUE TITLE
[a11y] Keyboard focus not visible on buttons in High Contrast

### ISSUE THREAD

**Issue by dpersing** [Shopify/polaris-react#794]
## Issue summary

The [Button](https://polaris.shopify.com/components/actions/button#navigation) component relies on changes in border color, background color, and/or fill color alone to convey focus states. This hides focus states from Windows High Contrast users, who only see the colors provided by their theme.

**Edit:** For the Plain button, focus appears to hide the element altogether, where the text is hidden on focus as well. This is caused by the `background` styles on the interior `<span>` `:after:`  becoming visible on focus, and being translated as solid instead of semi-transparent by HC. This issue does not affect icons implemented as part of Plain buttons.

![Standard button focus style](https://user-images.githubusercontent.com/1462085/41797438-dc74d794-761e-11e8-96a5-7a7789a06ec4.png)

![Basic button inactive, active, focus, and hover style in High Contrast](https://user-images.githubusercontent.com/1462085/44373386-c95df980-a49d-11e8-8b7d-2ff778b36f9b.png)

![Plain button focus style](https://user-images.githubusercontent.com/1462085/41797441-dcb25416-761e-11e8-9d3a-7167e1459f69.png)

![Plain button in High Contrast](https://user-images.githubusercontent.com/1462085/44373408-e0045080-a49d-11e8-9e72-660d8d86380c.png)

![Plain button focus in High Contrast](https://user-images.githubusercontent.com/1462085/44373419-e98db880-a49d-11e8-977a-c9b826395a23.png)

![Plain icon-only button focus style](https://user-images.githubusercontent.com/1462085/41797442-dcc6ea3e-761e-11e8-9680-301a81b01b00.png)

![Primary button focus style](https://user-images.githubusercontent.com/1462085/41797440-dc9cf3f0-761e-11e8-9a8d-2e1a7149c5ab.png)

![Destructive button focus style](https://user-images.githubusercontent.com/1462085/41797439-dc896380-761e-11e8-80f5-c526188716ec.png)

#### Standard button focus (CSS)

```
.Polaris-Button:focus {
    border-color: #5c6ac4;
    outline: 0;
    box-shadow: 0 0 0 1px #5c6ac4;
}
```

#### Plain button (CSS)

```
.Polaris-Button--plain {
    fill: #007ace;
    margin: -.7rem -.8rem;
    padding-left: .8rem;
    padding-right: .8rem;
    background: transparent;
    border: 0;
    box-shadow: none;
    color: #007ace;
}

.Polaris-Button--plain:active, .Polaris-Button--plain:focus, .Polaris-Button--plain:hover {
    fill: #084e8a;
    background: transparent;
    border: 0;
    box-shadow: none;
    color: #084e8a;
}
```

```
.Polaris-Button--plain>.Polaris-Button__Content:after {
    position: absolute;
    top: -2px;
    bottom: -2px;
    left: -5px;
    right: -5px;
    display: block;
    background: rgba(33,43,54,.1);
    border-radius: 3px;
    transition: opacity .2s cubic-bezier(.64,0,.35,1);
    content: "";
    opacity: 0;
}
```

#### Icon-only plain button (CSS)

```
.Polaris-Button--plain.Polaris-Button--iconOnly {
    fill: #637381;
    margin: -.8rem;
}

.Polaris-Button--plain.Polaris-Button--iconOnly:active, .Polaris-Button--plain.Polaris-Button--iconOnly:focus {
    fill: #637381;
    background: rgba(33,43,54,.1);
}
```

#### Primary button (CSS)

```
.Polaris-Button--primary {
    background: linear-gradient(180deg,#6371c7,#5563c1);
    border-color: #3f4eae;
    box-shadow: inset 0 1px 0 0 #6774c8, 0 1px 0 0 rgba(22,29,37,.05), 0 0 0 0 transparent;
    color: #fff;
    fill: #fff;
}

.Polaris-Button--primary:focus {
    border-color: #202e78;
    box-shadow: inset 0 1px 0 0 #6f7bcb, 0 1px 0 0 rgba(22,29,37,.05), 0 0 0 1px #202e78;
}
```

#### Destructive button (CSS)

```
.Polaris-Button--destructive {
    background: linear-gradient(180deg,#e6391a,#d53417);
    border-color: #b02b13;
    box-shadow: inset 0 1px 0 0 #e73d1f, 0 1px 0 0 rgba(22,29,37,.05), 0 0 0 0 transparent;
    color: #fff;
    fill: #fff;
}

.Polaris-Button--destructive:focus {
    border-color: #bf0711;
    box-shadow: inset 0 1px 0 0 #e84528, 0 1px 0 0 rgba(22,29,37,.05), 0 0 0 1px #bf0711;
}
```

## Steps to reproduce

1. Navigate to the [Button](https://polaris.shopify.com/components/actions/button#navigation) component examples, or to any other component that uses the Button.
1. Set the focus state in the web inspector, or <kbd>tab</kbd> to the button to see its focus state. Adjust the buttons to be text-only, text/icon, and icon-only to see different color combinations.
1. Turn on any [High Contrast theme in Windows settings](https://support.microsoft.com/en-us/help/13862/windows-use-high-contrast-mode) and repeat the process for each button style.

### Expected

- Keyboard focus will be visible.

### Actual

- Keyboard focus is not visible in High Contrast since the color-based styles are removed.
- When the plain button receives focus, its text is hidden because's it's hidden by an absolutely-positioned `:after` pseudo-element with a background style.

## Recommendations

- Allow Windows High Contrast users to see the native focus `outline` or a similar focus state not relying on color. The `-ms-high-contrast` media query can be used to show these states, but should not be used to assign specific colors since these may conflict with user settings.
- Refrain from using absolute positioning on `:after:` pseudo-elements to apply styles to elements, since they stack on "top" of the element. Instead, use CSS to change the background of the element itself.
- Consider doing usability testing with low vision users who use a variety of settings and tools to ensure the no-focus/focus difference is sufficient for ease of use.

## Specifications

- **Component:** Button, and components that leverage Button
- **WCAG SCs:** [1.4.1 Use of Color (Level A)](https://www.w3.org/TR/WCAG21/#use-of-color), [2.4.7 Focus Visible (Level AA)](https://www.w3.org/TR/WCAG21/#focus-visible)
- **Testing done with:**
 - Visual inspection
 - Windows High Contrast/Edge/Windows 10

**Comment by dpersing** [comment]
Thanks so much for taking a look at this @francinen! Let me know what I can do to help.
